charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_name.py

```diff
@@ -0,0 +1,121 @@
+LIBPATCH = 1
+
+
+DEFAULT_RELATION_NAME = "gateway"
+DEFAULT_INTERFACE_NAME = "istio-gateway-name"
+
+logger = logging.getLogger(__name__)
+
+
+class GatewayRelationError(Exception):
+    pass
+
+
+class GatewayRelationMissingError(GatewayRelationError):
+    def __init__(self):
+        self.message = "Missing gateway relation with istio-pilot"
+        super().__init__(self.message)
+
+
+class GatewayRelationTooManyError(GatewayRelationError):
+    def __init__(self):
+        self.message = "Too many istio-gateway-name relations"
+        super().__init__(self.message)
+
+
+class GatewayRelationDataMissingError(GatewayRelationError):
+    def __init__(self, message):
+        self.message = message
+        super().__init__(self.message)
+
+
+class GatewayRequirer(Object):
+    def __init__(self, charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+
+    def get_relation_data(self):
+        if not self.model.unit.is_leader():
+            return
+        gateway = self.model.relations[self.relation_name]
+        if len(gateway) == 0:
+            raise GatewayRelationMissingError()
+        if len(gateway) > 1:
+            raise GatewayRelationTooManyError()
+
+        remote_app = [
+            app
+            for app in gateway[0].data.keys()
+            if isinstance(app, Application) and not app._is_our_app
+        ][0]
+
+        data = gateway[0].data[remote_app]
+
+        if not "gateway_name" in data:
+            logger.error(
+                "Missing gateway name in gateway relation data. Waiting for gateway creation in istio-pilot"
+            )
+            raise GatewayRelationDataMissingError(
+                "Missing gateway name in gateway relation data. Waiting for gateway creation in istio-pilot"
+            )
+
+        if not "gateway_namespace" in data:
+            logger.error("Missing gateway namespace in gateway relation data")
+            raise GatewayRelationDataMissingError(
+                "Missing gateway namespace in gateway relation data"
+            )
+
+        return {
+            "gateway_name": data["gateway_name"],
+            "gateway_namespace": data["gateway_namespace"],
+        }
+
+
+class GatewayProvider(Object):
+    def __init__(self, charm, lightkube_client, resource_handler):
+        super().__init__(charm, DEFAULT_RELATION_NAME)
+        self.lightkube_client = lightkube_client
+        self.charm = charm
+        self.resource_handler = resource_handler
+        self.framework.observe(
+            charm.on[DEFAULT_RELATION_NAME].relation_changed, self._on_gateway_relation_changed
+        )
+        self.framework.observe(charm.on.config_changed, self._on_gateway_config_changed)
+        self.framework.observe(charm.on.update_status, self._on_gateway_config_changed)
+
+    def _validate_gateway_exists(self):
+        response = self.lightkube_client.get(
```

Comment:
I'd add a try/catch block here, this `lightkube_client.get()` could end up raising errors that will prevent this response variable to be defined. 
Also, thinking about this method a bit more, `lightkube_client.get()` returns an object that belongs to a class, in this case the class of the resource you are looking for, `Gateway`. You could do something very redundant, but also very explicit (since this is a library, it should be okay) like the following:

```python
from lightkube.generic_resource import create_namespaced_resource
...
def __init__(...):
  self.gateway_class = create_namespaced_resources("define the Gateway resource here")

def _validate_gateway_exists(self):
  try:
    gateway_resource = client.get(self.gateway_class, self.model.config['default-gateway'], namespace=self.model.name)
  except ApiError:
   some_error_handling
  if isinstance(gateway_resource, gateway_class):
    return True
  return False
```
This helps you to catch any Exceptions the lightkube client might throw and also avoids the dependency on resource_handler.
If you don't want to do a redundant check, you can leave the `get` and return a True if no exceptions are raised, False otherwise.